src/arch/CMakeLists.txt

```diff
@@ -9,4 +9,9 @@ elseif(HAVE_PPC64LE OR HAVE_PPC64 OR HAVE_PPC)
   list(APPEND arch_srcs ppc.c)
 endif()
 
-add_library(arch STATIC ${arch_srcs})
+if(WIN32)
+  # We need to avoid duplicate symbols with llvm.
+  add_library(arch OBJECT ${arch_srcs})
+else()
+  add_library(arch STATIC ${arch_srcs})
+endif()
```

Comment:
this change further segments the building system. it hurts the maintainability. please explain which libraries contain `arch` and why they duplicate. for instance what shared library or executable includes these two libraries.

if that's the case, we should apply this change to non-win32 builds.

Reply:
The issue with ``crc32`` was that some symbols were re-exposed by libceph-common, hitting linker errors when linking against libceph-common _and_ the static crc32 lib:

```
FAILED: bin/rbd.exe
: && /mnt/data/workspace/ceph_mingw_clang/build.deps/mingw-llvm/bin/x86_64-w64-mingw32-clang++ -g1 -O3 -DNDEBUG  src/common/CMakeFiles/common_texttable_obj.dir/TextTable.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/rbd.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/ArgumentTypes.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/IndentStream.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/MirrorDaemonServiceInfo.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/OptionPrinter.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/Schedule.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/Shell.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/Utils.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Bench.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Children.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Clone.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Config.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Copy.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Create.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Device.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Diff.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/DiskUsage.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Encryption.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Export.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Feature.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Flatten.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Ggate.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Group.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/ImageMeta.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Import.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Info.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Journal.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Kernel.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/List.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Lock.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/MergeDiff.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Migration.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/MirrorImage.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/MirrorPool.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/MirrorSnapshotSchedule.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Namespace.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Nbd.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/ObjectMap.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Perf.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/PersistentCache.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Pool.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Remove.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Rename.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Resize.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Snap.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Sparsify.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Status.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/TrashPurgeSchedule.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Trash.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Ubbd.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Watch.cc.obj src/tools/rbd/CMakeFiles/rbd.dir/action/Wnbd.cc.obj -o bin/rbd.exe -Wl,--major-image-version,0,--minor-image-version,0 -L/mnt/data/workspace/ceph_mingw_clang/build.deps/mingw/zlib/lib lib/libcls_journal_client.a  lib/libcls_rbd_client.a  lib/librbd_types.a  lib/liblibrbd.a  lib/libjournal.a  lib/liblibneorados.a  lib/liblibrados.a  lib/libglobal.a  lib/libcls_rbd_client.a  lib/libcls_journal_client.a  lib/liblibrados_impl.a  lib/libcls_lock_client.a  lib/librbd_internal.a  lib/librbd_types.a  lib/libosdc.a  ../build.deps/mingw/openssl/lib/libssl.dll.a  -lpthread  lib/libceph-common.dll.a  ../build.deps/mingw/openssl/lib/libcrypto.dll.a  lib/libjson_spirit.a  lib/libcommon_utf8.a  lib/liberasure_code.a  lib/libextblkdev.a  lib/libcrc32.a  ../build.deps/mingw/boost/lib/libboost_thread.lib  ../build.deps/mingw/boost/lib/libboost_chrono.lib  ../build.deps/mingw/boost/lib/libboost_system.lib  ../build.deps/mingw/boost/lib/libboost_random.lib  ../build.deps/mingw/boost/lib/libboost_program_options.lib  ../build.deps/mingw/boost/lib/libboost_date_time.lib  ../build.deps/mingw/boost/lib/libboost_iostreams.lib  -lz  ../build.deps/mingw/zlib/lib/libz.dll.a  ../build.deps/mingw/libbacktrace/lib/libbacktrace.a  -lpthread  -lws2_32  -lmswsock  -liphlpapi  -lbcrypt  lib/libdlfcn_win32.a  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 && :
ld.lld: error: ceph_crc32c_zeros was replaced
```

Same with the ``arch`` lib:

```
FAILED: bin/unittest_arch.exe
: && /mnt/data/workspace/ceph_mingw_clang/build.deps/mingw-llvm/bin/x86_64-w64-mingw32-clang++ -g1 -O3 -DNDEBUG  src/test/CMakeFiles/unit-main.dir/unit.cc.obj src/test/CMakeFiles/unittest_arch.dir/test_arch.cc.obj -o bin/unittest_arch.exe -Wl,--major-image-version,0,--minor-image-version,0 -L/mnt/data/workspace/ceph_mingw_clang/build.deps/mingw/zlib/lib lib/libgmock_main.a  lib/libgmock.a  lib/libgtest.a  -lpthread  -lws2_32  lib/libglobal.a  lib/libceph-common.dll.a  lib/libjson_spirit.a  lib/libcommon_utf8.a  lib/liberasure_code.a  lib/libextblkdev.a  lib/libarch.a  ../build.deps/mingw/boost/lib/libboost_thread.lib  ../build.deps/mingw/boost/lib/libboost_chrono.lib  ../build.deps/mingw/boost/lib/libboost_system.lib  ../build.deps/mingw/boost/lib/libboost_random.lib  ../build.deps/mingw/boost/lib/libboost_program_options.lib  ../build.deps/mingw/boost/lib/libboost_date_time.lib  ../build.deps/mingw/boost/lib/libboost_iostreams.lib  -lz  ../build.deps/mingw/zlib/lib/libz.dll.a  ../build.deps/mingw/libbacktrace/lib/libbacktrace.a  -lpthread  -lws2_32  ../build.deps/mingw/openssl/lib/libcrypto.dll.a  -lmswsock  -liphlpapi  -lbcrypt  lib/libdlfcn_win32.a  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 && :

ld.lld: error: ceph_arch_probe was replaced
clang-16: error: linker command failed with exit code 1
```

I've updated the commit to use ``OBJECT`` on all platforms.

Reply:
After switching to ``OBJECT`` on all platforms, the linux job fails with this:

```
/usr/bin/ld: /usr/bin/ld: DWARF error: invalid or unhandled FORM value: 0x23
lib/libcrimson-common.a(crc32c.cc.o): in function `ceph_choose_crc32':
crc32c.cc:(.text+0x9): undefined reference to `ceph_arch_probe'

There's probably an unspecified dependency.
```

Reply:
i think the crux of the problem is that 

> The issue with `crc32` was that some symbols were re-exposed by libceph-common, hitting linker errors when linking against libceph-common _and_ the static crc32 lib:
> 

how about linking `crc32` against `ceph-common` instead of `arch`

Reply:
Thanks for the suggestion but that wouldn't work either as ``ceph-common`` needs ``crc32`` functions:

```
ld.lld: error: undefined symbol: ceph_crc32c_func
>>> referenced by src/msg/CMakeFiles/common-msg-objs.dir/Message.cc.obj:(.refptr.ceph_crc32c_func)
>>> referenced by src/msg/CMakeFiles/common-msg-objs.dir/async/ProtocolV1.cc.obj
>>> referenced by src/msg/CMakeFiles/common-msg-objs.dir/async/frames_v2.cc.obj

ld.lld: error: undefined symbol: ceph_crc32c_zeros
>>> referenced by /mnt/data/workspace/ceph_mingw_clang/src/include/crc32c.h:47
>>>               src/msg/CMakeFiles/common-msg-objs.dir/async/frames_v2.cc.obj:(ceph::msgr::v2::FrameAssembler::disassemble_preamble(ceph::buffer::v15_2_0::list&))
>>> referenced by /mnt/data/workspace/ceph_mingw_clang/src/include/crc32c.h:47
>>>               src/common/CMakeFiles/common_buffer_obj.dir/buffer.cc.obj:(ceph::buffer::v15_2_0::list::crc32c(unsigned int) const)
>>> referenced by /mnt/data/workspace/ceph_mingw_clang/src/include/crc32c.h:47
>>>               src/common/CMakeFiles/common_buffer_obj.dir/buffer.cc.obj:(ceph::buffer::v15_2_0::list::crc32c(unsigned int) const)
>>> referenced 2 more times
clang-16: error: linker command failed with exit code 1 (use -v to see invocation)
```

Reply:
i mean, now that

0. `crc32` depends on `arch`, and
1. `crc32` ends up being included by `common`:  crc32 => ceph_common_deps => common, and some objects in common indeed depend on it, and
2.  `common` includes `arch` , as you pointed out.

so i am wondering if the right direction is to have following dependency:

common = link public library => crc32 => link public library => arch

and let crc32c executable to use common, so it will have access to crc32 via common. 

what do you think?

Reply:
Tried ``target_link_libraries(crc32 PUBLIC arch)`` but I was still getting the following:

```
[2/2] Linking CXX executable bin/unittest-crimson-backfill
FAILED: bin/unittest-crimson-backfill
...
/mnt/data/workspace/ceph_linux/src/common/crc32c.cc:21: undefined reference to `ceph_arch_probe'
/usr/bin/ld: /mnt/data/workspace/ceph_linux/src/common/crc32c.cc:26: undefined reference to `ceph_arch_intel_sse42'
collect2: error: ld returned 1 exit status
```

Linking ``crimson`` directly against ``arch`` seems to work fine, I've added a commit that can be squashed afterwards.